### PR TITLE
feat(bundler) code signing

### PR DESCRIPTION
**Pull request by lucasfernog**
- [ ] Bugfix
- [x] Feature
- [ ] New Binding Issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes. Issue #___
- [x] No

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch and _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

I'm opening this PR because I don't fully understand the scope, so if you know something about code signing that's not yet implemented here, please comment here.

### CODE REVIEW

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (rajivshah3) [inline cli/tauri-bundler/src/bundle/osx_bundle.rs]: Can we show the `stderr` output in case signing fails? That should make it easier for users to figure out what's going wrong

Comment (lucasfernog) [reply cli/tauri-bundler/src/bundle/osx_bundle.rs]: yeah that makes sense, gonna do it

**Comment by nothingismagick** [comment]
maybe we revisit this after #643 lands.

**Comment by lucasfernog** [comment]
I'll be able to revisit next week.

**Comment by olegshilov** [comment]
Hello comrades.
Is there any news about this request?

**Comment by lucasfernog** [comment]
Unfortunately I can't test this anymore without an enrollment with the Apple Developer Program :/

**Comment by nothingismagick** [comment]
@rajivshah3 - I know you were recently able to do this in CI. Can you confirm it is working here?

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (rajivshah3) [inline cli/tauri-bundler/src/bundle/osx_bundle.rs]:
```suggestion
    identity,
    "--options",
    "runtime"
```

This is required in order to notarize an app:

```
runtime  On macOS versions >= 10.14.0, opts signed processes into a hard-
              ened runtime environment which includes runtime code signing
              enforcement, library validation, hard, kill, and debugging
              restrictions.  These restrictions can be selectively relaxed via
              entitlements. Note: macOS versions older than 10.14.0 ignore the
              presence of this flag in the code signature.
```

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (rajivshah3) [inline cli/tauri-bundler/src/bundle/osx_bundle.rs]: It might be easier to parse this if you pass `--output-format xml` or `--output-format json` to `xcrun altool`

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (rajivshah3) [inline cli/tauri-bundler/src/bundle/osx_bundle.rs]: This is pretty rare, but if you're a member of multiple teams in App Store Connect you need to specify which team you're uploading for. You can pass `--asc-provider` (see https://developer.apple.com/documentation/xcode/notarizing_macos_software_before_distribution/customizing_the_notarization_workflow)

**Review by rajivshah3 — COMMENTED** [review]
Just a couple of small things based on problems I/others have run into with notarization

**Comment by rajivshah3** [comment]
@nothingismagick sure, I think after https://github.com/tauri-apps/tauri/pull/473#discussion_r536915775 is addressed (I think Apple will reject it otherwise) then I can test it out

**Comment by lucasfernog** [comment]
@rajivshah3 I'd love you even more if you could see if the notarization is working, I had access to an apple developer program earlier this year but I lost it :(  I can make whatever change needed.

**Comment by rajivshah3** [comment]
@lucasfernog sure, I can test it out now that the `--options runtime` is added. Do you want to resolve the merge conflicts first or should I test it now?

**Comment by lucasfernog** [comment]
@rajivshah3 merge done, I'll fix windows later

oh boy I was terrible with Rust, I can cleanup this code later this week.

File: cli/tauri-bundler/src/bundle/common.rs
Comment (nklayman) [inline cli/tauri-bundler/src/bundle/common.rs:280]: Why move this out of the main import statement?

File: cli/tauri-bundler/src/bundle/common.rs
Comment (nklayman) [inline cli/tauri-bundler/src/bundle/common.rs:293]: Same as above

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (nklayman) [inline cli/tauri-bundler/src/bundle/osx_bundle.rs]: I think it would be a good idea to move all this logic into a separate file as this one is quite long and complicated now.

File: cli/tauri-bundler/src/bundle/common.rs
Comment (lucasfernog) [reply cli/tauri-bundler/src/bundle/common.rs:280]: because it doesn't run on Windows, there's something wrong happening and i'll take a look. (and unused imports are a clippy warning).

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (lucasfernog) [reply cli/tauri-bundler/src/bundle/osx_bundle.rs]: yeah I agree with you, we should do it when this is properly working :D

File: cli/tauri-bundler/src/bundle/settings.rs
Comment (nklayman) [inline cli/tauri-bundler/src/bundle/settings.rs]: Why are the windows attributes conditionally compiled but the macos ones aren't?

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (rajivshah3) [inline cli/tauri-bundler/src/bundle/osx_bundle.rs]: Are `--deep` and `--force` necessary here?

image

https://developer.apple.com/library/archive/technotes/tn2206/_index.html#//apple_ref/doc/uid/DTS40007919-CH1-TNTAG404

File: cli/tauri-bundler/src/bundle/wix.rs
Comment (nklayman) [inline cli/tauri-bundler/src/bundle/wix.rs:561]: Just like with MacOS, I think we should move this to a separate file for simplicity and to avoid a nearly 1000-line file.

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (rajivshah3) [inline cli/tauri-bundler/src/bundle/osx_bundle.rs]: Do we need to sign depth-first? `electron-osx-sign` recently began doing this:

https://github.com/electron/electron-osx-sign/blob/9bc2eedae5ece7a15576e1d564ed0ddb67a9e110/sign.js#L201-L210

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (rajivshah3) [inline cli/tauri-bundler/src/bundle/osx_bundle.rs]:
```suggestion
        common::print_info(format!("skipping app notarization, {}", e.to_string()).as_str())?;
```

**Review by nklayman — CHANGES_REQUESTED** [review]
I haven't tested it, but it looks like it would work :joy:. I just think we should move the signing logic into separate files since the build files are so long now.

File: cli/tauri-bundler/src/bundle/settings.rs
Comment (lucasfernog) [reply cli/tauri-bundler/src/bundle/settings.rs]: because the windows stuff uses some APIs that won't compile on unix systems, so everything is conditionally compiled.

File: cli/tauri-bundler/src/bundle/wix.rs
Comment (lucasfernog) [reply cli/tauri-bundler/src/bundle/wix.rs:561]: > I haven't tested it, but it looks like it would work 😂. I just think we should move the signing logic into separate files since the build files are so long now.

it doesn't work lol i've tested it last year.

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (lucasfernog) [reply cli/tauri-bundler/src/bundle/osx_bundle.rs]: yeah i think we do, I wanted to do it but I can't test it :(

File: cli/tauri-bundler/src/bundle/common.rs
Comment (nklayman) [reply cli/tauri-bundler/src/bundle/common.rs:280]: Ahh that part is non-windows only, my bad.

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (rajivshah3) [reply cli/tauri-bundler/src/bundle/osx_bundle.rs]: Right. I can test it for you. I'm actually not sure whether that depth-first thing was specific to Electron tbh 🤔  but I can try it out

Comment (lucasfernog) [reply cli/tauri-bundler/src/bundle/osx_bundle.rs]: it's better to do it manually? :(

Comment (lucasfernog) [reply cli/tauri-bundler/src/bundle/osx_bundle.rs]: probably not :D I know something about Apple :joy: 

thanks!

Comment (rajivshah3) [reply cli/tauri-bundler/src/bundle/osx_bundle.rs]: I'm not entirely sure why 😄  I just know Apple really doesn't recommend it https://developer.apple.com/forums/thread/128166
https://developer.apple.com/forums/thread/129980

Comment (lucasfernog) [reply cli/tauri-bundler/src/bundle/osx_bundle.rs]: want me to change it? I'd do anything to get this PR done :D

Comment (rajivshah3) [reply cli/tauri-bundler/src/bundle/osx_bundle.rs]: Yeah, I don't think those are necessary unless we find that it doesn't work without `--deep` and `--force`. The fact that an Apple employee wrote an article called [`--deep` Considered Harmful](https://developer.apple.com/forums/thread/129980) sounds really concerning 😂

Comment (lucasfernog) [reply cli/tauri-bundler/src/bundle/osx_bundle.rs]: absolutely
I thought --deep was necessary to sign all files
and --force I think I was just using because I was testing signing on a previously signed app :joy:

File: cli/tauri-bundler/src/bundle/osx_bundle.rs
Comment (lucasfernog) [reply cli/tauri-bundler/src/bundle/osx_bundle.rs]: I think i'll restructure the bundler crate on a separate PR.

**Comment by lucasfernog** [comment]
@rajivshah3 i'll merge this but let me know if there's any change we need to address.

**Pull request merged**
